Changes made in preparation for the XCSP3 and MiniZinc competitions

cprudhom opened the pull request:
These changes include the following:
- Change to the default strategy for XCSP3
- Conditional transformation of expressions into constraints in extensions and detection of some classic expressions (see #embody() method)
- Implementation of a cache for the XCSP3 parser
- addition of a configurable RoundRobin strategy
- improvement and correction of DomOverWDeg and its variants
- improvement of structures for CompactTable
- creation of a class dedicated to preprocessing algorithms
- addition of flexibility in ParallelPortfolio
- other improvements and bug fixes

cprudhom committed:
Add decomposition of AllDifferentExcept0 and AllDifferentExceptValues in XCSP (+ to fzn)

cprudhom committed:
Fix error in XCSP groups declaration.
The degree of the XCSP variable should be different from 0 to be really created in the choco model.

cprudhom committed:
Reduce the extension fallback threshold to 100_000 instead of 2147483

cprudhom committed:
Add new lzma files

cprudhom committed:
Add seed as new parameter in PickOnDom and PickOnFil

cprudhom committed:
Update BlackBoxConfigurator#setRefinedPartialAssignmentGeneration to return this

cprudhom committed:
Add MetaStrategy abstract class

cprudhom committed:
Remove PropSat from failure-based search strategy because the "explanation" of a failure is too big (=all variables)

cprudhom committed:
Fix error when getting the group of variables and some are null (XCSP3)

cprudhom committed:
Change conditions for choosing between extension or decomposition for Expression (XCSP3)

cprudhom committed:
Simplify PropSTR2 to avoid storing the Tuples object

cprudhom committed:
Cache matrix (using SoftReference) in the Tuples object to reduce memory footprint when using PropSTR2+

cprudhom committed:
Cache tuples (using WeakHashMap) in XCSPParser to reduce memory footprint

cprudhom committed:
Add a RoundRobin strategy based on Group

cprudhom committed:
Remove assert k > 0 for LastConflict to allow dummy meta strategy declaration (for RoundRobin mainly)

cprudhom committed:
AbstractCriterionBasedVariableSelector now interpert seed = -1 as lexico order for breaking ties

cprudhom committed:
AbstractCriterionBasedVariableSelector now interpert seed = -1 as lexico order for breaking ties

cprudhom committed:
Add NextModuloN policy for MAB (= round robin)

cprudhom committed:
Update Search#roundRobinSearch and create Search#adaptiveRoundRobinSearch

cprudhom committed:
Fix java compatibility

cprudhom committed:
Define a default search strategy for XCSP3

jgFages committed:
support regular with negative values

jgFages committed:
faster isInScope method + bugfix conflictOrderingSearch

+add model and decision path to AbstractStrategy

jgFages committed:
search simplification

jgFages committed:
update changes.md

jgFages committed:
fix tests

jgFages committed:
fix test

cprudhom committed:
Add Javadoc

jgFages committed:
support regular with negative values

cprudhom committed:
Move MapVal and ArrayVal to IVal inner classes.
+ delete ValueSortedMap.java

cprudhom committed:
Create Weightable class and its two inner classes.
Can be used to map double to Identity-based objects

cprudhom committed:
Correction of ConflictHistorySearch

cprudhom committed:
Revamp AbstractCriterionBasedVariableSelector.
- Replace TObjectDoubleMap by Weightable for efficiency reason
- Dom/WDeg now manages futvars (via Element) for itself and Dom/WDegRef
- remove AbstractCriterionBasedVariableSelector#increase (only required by Dom/WDeg and Dom/WDegRef)

cprudhom committed:
Fix https://github.com/chocoteam/choco-solver/pull/1154#discussion_r2017726850 and https://github.com/chocoteam/choco-solver/pull/1154#discussion_r2017727171

cprudhom committed:
Remove PickOnFil.java

cprudhom committed:
Adapt the philosophy of PickOnDom to constraint-oriented prop engine. The variables enqueued during each fixpoint that ends with failure are collected and their weights are update with the value `(l - i) / l` where l is the number of enqueued variables and i the position of the variable in the list.
- reduce variants to one
- remove PropagationInsight

cprudhom committed:
Remove Weightable and replace the Array implementation by IdentityToDouble

cprudhom committed:
Move IVal to inner class of ActivityBased and change its name to ValueActivity.
+ Change threshold to switch between Array and Map
+ remove call to `ensureCapacity` from `get`

cprudhom committed:
Cache transitions for MDDs in XCSP3

cprudhom committed:
Change priority of STR2+

cprudhom committed:
Remove automatic conversion of scalar to table

cprudhom committed:
Better (correct) conditions for propagating half-reified constraint

cprudhom committed:
Fix error in allDifferentExceptValues

cprudhom committed:
Fix issue related to `in`

cprudhom committed:
Fix filtering weakness in atLeastNValues

cprudhom committed:
Check view by default

cprudhom committed:
Model is not final anymore

cprudhom committed:
`abs` (fake) view does not force the type of domain of the resulting var

cprudhom committed:
Fix error when using range with views

cprudhom committed:
Does not consider constants as variables when collecting variables from expressions

cprudhom committed:
Recognize x.dist(y).eq(z) from expression like x.sub(y).abs().eq(z)

cprudhom committed:
Improvement of CT:
- performance amelioration
- create CT for negative table

cprudhom committed:
Replace TIntSet with BitSet in PropLargeMDDC

cprudhom committed:
Change the way STAR value is set in Tuples

cprudhom committed:
Test non uniqueness of a variable in table constraint

cprudhom committed:
Add methods in TuplesFactory to sample positive tuples based on variables and TupleValidator to estimate whether or not the table should be post on positive tuples or negative tuples.
Add also a method to return both positive and negative tuples

cprudhom committed:
Correct IIntDeltaMonitor#sizeApproximation()

cprudhom committed:
CompactTable: sparse domains are now managed with dedicated support structure to reduce memory footprint

cprudhom committed:
Fix error in T_solve_goal#testNestedSeq model definition (related to #0602a4e0)

cprudhom committed:
Correct wrong domain definition for abs expression

cprudhom committed:
Downgrade slf4j-nop version to remove message with XCSP

cprudhom committed:
Update Makefile with ISO 8601 format for Welcome message in XCSP and Flatzinc parsers

cprudhom committed:
Remove SatDecorator#removeDominated(cr) call on learnClause(...) because it is too costly with regard to the gain obtained

cprudhom committed:
NogoodFromRestarts better handle BoolVar

cprudhom committed:
All implementations of ArExpression override #getExpressionChild()
+ add #getMaxDepth() API for expressions

cprudhom committed:
Passing -1 as a seed to FailureBased search breaks tie with input order

cprudhom committed:
Fix the way FutVars are managed in Dom/Wdeg and refined

cprudhom committed:
Operator of LoExpression are accessible

cprudhom committed:
Fix issue related to IntVarLazyLit and getLit(v,LR_EQ)

cprudhom committed:
Remove IntDomainRandomBound.java and add new parameter to IntDomainRandom to force random selection over bounds only

cprudhom committed:
Fix bug related to wrong conversion to long in PropTimesNaiveWithLong

cprudhom committed:
Improve Clause#reduceOs

cprudhom committed:
Bind r(int... ps) to r(int) and r(int, int)

cprudhom committed:
Simplification of BoolEvtScheduler

cprudhom committed:
Fix incorrect management of reification when views are implied

cprudhom committed:
Fix incorrect creation of univ value for IfThenElse

cprudhom committed:
Update Docker.dms

cprudhom committed:
Update the docker command of Makefile to generate two images, one with lcg and one with

cprudhom committed:
Change parameters of RegParser

cprudhom committed:
Add decomposition of Disjunctive

cprudhom committed:
Turn scalar into table in the decomposition of 2D-Element

cprudhom committed:
Declare third-party licenses in uberjar/META-INF

cprudhom committed:
Update maven configuration to central sonatype prerequisites

cprudhom committed:
Fix raw uses of Decision

cprudhom committed:
Deal with NE operator in NaReExpression

cprudhom committed:
UnArExpression: disable distance recognition when LCG is on

cprudhom committed:
LCG: dedicated propagators for div and abs

cprudhom committed:
Table: simplify default algo selection, now CT is the unique falling choice

cprudhom committed:
Preprocessing: move the class and add new actions

cprudhom committed:
Flatzinc : clean up cumulative decomp option

cprudhom committed:
XCSP3: adapt restart and nogood recording to LCG option

cprudhom committed:
BBox search conf: use 'restartOnSolution' option

cprudhom committed:
Bbox searches: set flush threshold to 20 restarts by default + allow disabling random tie breaking

cprudhom committed:
MetaStrategy: declare abstract class for more flexibility

cprudhom committed:
IntValueSelection: add sticky

cprudhom committed:
Search: update RoundRobin

cprudhom committed:
Expression: add 'embody' method that decides whether to decompose the expression as a set of constraints or to represent it as an extensional constraint

cprudhom committed:
ParallelPortfolio : delete 'searchAutoConf' parameter, now each model, when added, indicates whether its strategy should be kept or overridden

cprudhom committed:
XCSP3: update parsing

cprudhom committed:
int_div: add missing (and obvious) filtering for the light version

cprudhom committed:
at_least_nvalues: rollback #5b8d7ab4 (incorrect)

cprudhom committed:
int_div: capture unsatisfiable cases

cprudhom committed:
Tests: update tests for explanations

cprudhom committed:
XCSP3: move to xcsp3-tools v2.5

cprudhom committed:
Fix last commits

cprudhom committed:
abs: better deal with constants

cprudhom committed:
div: catch unsupported cases of light version and fall back to full version

cprudhom committed:
fix tests wrt to commit #3f947a97 which change the behavior of the default search

cprudhom committed:
fix commit #c5d5c4c1 : accidental change in the scope of the variable currentVarIndex

cprudhom committed:
revert commit #4f69c42b (not justified enough)

cprudhom committed:
Table: adapt priority of CT wrt the number of variables involved

cprudhom committed:
Expr : fix seed in sampling for conversion to extensional constraint

cprudhom committed:
XCCSP: update tests wrt to changes related to competition preparation

cprudhom committed:
Table: fix test wrt commuit #e86270e7

jgFages commented on solver/src/main/java/org/chocosolver/solver/constraints/IDecompositionFactory.java:
```diff
@@ -46,6 +48,55 @@
+     * @param vars   array of integer variables
+     * @param except array of values that can be excluded from the allDifferent constraint
+     */
+    default void allDifferentExceptDec2(IntVar[] vars, int[] except) {
```
Maybe rename as "allDifferentExceptGCCDec" ?

cprudhom replied:
Why GCC?
allDifferentExceptManyDec ?

jgFages replied:
because the decomposition is based on GCC, so I would understand it as a "GCC based decomposition"
but I admit that reading "ExceptGCC" might be confusing, so perhaps allDifferentExceptDecGCC would be better.
We can also stay with the allDifferentExceptDec2

ArthurGodet replied:
What about "allDifferentExceptDecBasedOnGCC" ? Or "allDifferentExceptDecByGCC" ?

jgFages commented on solver/src/main/java/org/chocosolver/solver/constraints/reification/PropXgeCHalfReif.java:
```diff
@@ -46,7 +46,7 @@ public int getPropagationConditions(int vIdx) {
             return IntEventType.DECUPP.getMask();
         }
-        return IntEventType.INCLOW.getMask();
+        return IntEventType.INSTANTIATE.getMask();
```
change made to trigger the setPassive ?

cprudhom replied:
exactly

jgFages commented on solver/src/main/java/org/chocosolver/solver/search/strategy/selectors/values/IntDomainSticky.java:
```diff
@@ -0,0 +1,106 @@
+ *
+ * @author Charles Prud'homme
+ */
+public final class IntDomainSticky implements IntValueSelector, Function<IntVar, OptionalInt>, IVariableMonitor<IntVar> {
```
not used in round robin?

cprudhom replied:
Should be. I'll change that

cprudhom replied:
Actually, it is already usable in RoundRobin

cprudhom committed:
Merge branch 'develop' into xcsp

cprudhom committed:
fix commit #280c5848 (merge manually on github)

cprudhom committed:
LCG: prevent prohibiting-solution clauses from being removed on calls to MiniSat#doReduceDB().
This is done by storing separately failure clauses and prohibiting-solution ones in MiniSat#learnts

cprudhom committed:
LCG: fix a bug that occurs when some constraints are reified. Some events generated by propagateSat() can be lost because of `onPropagatorExecution` not placed correctly
+ fix typo in IntCircularQueue#toString

cprudhom committed:
CPProfiler: display modifications

cprudhom committed:
Flatzinc: set complementary search by default (not only output variables). See 2025/hitori with 5.1.dzn

cprudhom committed:
X+Y=Z: removal of `PropXplusYeqZ`, fall back propagator is `PropSum`

See https://gist.github.com/cprudhom/b94a841728ccf9ee24b2efcfb683115c for details

cprudhom committed:
Table: when the memory estimation is too big STR2+ is selected instead of CT
+ add new option 'maxSizeInMBToUseCompactTable' in Settings
+ STR2+ now manages tuples without ISet (IStateInt + int[] instead)

cprudhom committed:
LCG : constants are now managed as a specific case of IntVarEagerLit

cprudhom committed:
LCG : add new assertion when enqueuing literal in MiniSat
+ correction of few propagators which were incorrect
+ correction of false literal declaration (init of MiniSat)
+ correction of assertion message and showReason

jgFages commented on solver/src/main/java/org/chocosolver/sat/Clause.java:
```diff
@@ -139,6 +139,22 @@ public boolean learnt() {
+    /**
+     * Indicates if all literals but one are false in the clause.
+     * By implication, the remaining literal is the asserting literal.
+     * By contract, the literal at index 0 is the asserting literal and should not be considered.
```
I find the combination of "literal at index 0 ... should not be considered" and "allFalseButOne" confusing. 
My first understanding is that were are checking that all literals but one are false, starting from position 1, whereas the code check that they are all false. If the code is correct, I would rephrase the doc : 
"By contract, the literal at index 0 is the asserting literal and always true."

cprudhom replied:
Your first understanding is correct: we are checking that all literals but one are false, starting from position 1.
The loop starts at 1 (not 0) and if a literal is not false, the loop is broken and `false` is returned. 

Then, the contract is not that the literal at 0 is always true. 
In its current usage (for assertion only), it should be undefined, otherwise there is a problem somewhere (indeed, the current cannot explain an event that already occurs).

Note that if this method is called somewhere else in the code, the assertion can be wrong, though, since it must be true only when a literal is satisfied or falsified.

Its easier (from a conception perspective) to keep the method in the Clause class, but I agree that it can be renamed for clarity.

cprudhom committed:
MiniSat: move Clause#allFalseButOne to MiniSat#isAssertingClause

ArthurGodet commented on solver/src/main/java/org/chocosolver/sat/MiniSat.java:
```diff
@@ -659,7 +680,7 @@ private void analyseConflict(Clause confl, TIntList out_learnt, int p, int pathC
                 if (!seen.get(x) && level(x) > rootlvl) {
                     assert p == litUndef || pos(var(p)) > pos(x) : "chronological inconsistency :(" + printLit(p) + " @ " + pos(var(p)) +
-                            ") is explained by a previous event (" + printLit(x) + " @ " + pos(x) + ") " + c;
+                            ") is explained by a older event (" + printLit(x) + " @ " + pos(x) + ") " + c;
```
typo: "a older" instead of "an older"

ArthurGodet commented on solver/src/main/java/org/chocosolver/solver/constraints/IDecompositionFactory.java:
```diff
@@ -187,6 +238,31 @@ default void cumulativeDec(IntVar[] starts, IntVar[] durations, IntVar[] heights
+     * If a task is optional, it can be performed with a duration of 0 and the other mandatory tasks can overlap it.
+     * Otherwise, the task must be performed and cannot be overlapped by any other task even if it has a duration of 0.
+     *
+     * @param tasks array of tasks
```
the `@param strict` is missing

ArthurGodet commented on solver/src/main/java/org/chocosolver/solver/constraints/IDecompositionFactory.java:
```diff
@@ -187,6 +238,31 @@ default void cumulativeDec(IntVar[] starts, IntVar[] durations, IntVar[] heights
+     * Creates a decomposition of the disjunctive constraint.
+     * This constraint ensures that no two tasks overlap.
+     * <p>
+     * If a task is optional, it can be performed with a duration of 0 and the other mandatory tasks can overlap it.
+     * Otherwise, the task must be performed and cannot be overlapped by any other task even if it has a duration of 0.
```
This is not clear to me. I don't see, from the code, where the optionality of a task does intervene ? And shouldn't the optional tasks that are performed be considered as mandatory tasks ?

cprudhom replied:
Following the MiniZinc definition (for which this decomposition has been written) a task with 0-duration is considered as optional, and can be overlapped.
But, if strict is set to true, a 0-duration task cannot be overlapped.

ArthurGodet replied:
Ok this is clearer for the definition and I agree on the documentation.

I still would precise that optionality is treated as the possibility of the task to have a null-duration (as it is the only criteria taken into account in the code wrt optionality of tasks)

ArthurGodet commented on solver/src/main/java/org/chocosolver/solver/constraints/IIntConstraintFactory.java:
```diff
@@ -895,9 +939,34 @@ default Constraint allDifferentUnderCondition(IntVar[] vars, Condition condition
 
+    /**
+     * Creates an allDifferent constraint for variables that are not equal to 0.
+     * There can be multiple variables equal to 0.
```
Should not it be values instead of 0 ?

ArthurGodet commented on solver/src/main/java/org/chocosolver/solver/constraints/binary/PropAbsoluteLight.java:
```diff
@@ -0,0 +1,104 @@
+            int u = Y.getUB();
+            if (l >= 0) {
+                absY.updateLowerBound(l, this, lcg() ? Reason.r(Y.getMinLit()) : Reason.undef());
+                absY.updateUpperBound(u, this, lcg() ? Reason.r(Y.getMinLit(), Y.getMaxLit()) : Reason.undef());
```
Why is there `Y.getMinLit()` in the explanation ?

cprudhom replied:
Because l (Y.getLB()) is >=0, which could happen due to filtering of Y lower bound.

Consider the case where Y = [-10, 5], then |Y| = [0,10]
If an event set Y to [2,5], then |Y| = [2,5] and the modification of |Y| upper bound is due to the current upper bound of Y and the fact the lower bound of Y changes (to a value >= 0)

ArthurGodet replied:
I would have expected the explanation to therefore be `Reason.r(Y.getLELit(-1), Y.getMaxLit())` but it would only work for enumerated-domains' variables. Maybe we could rely on the `getNegGeqLit` and `getNegLeqLit` functions I used in PropagatorCumulative (by making them static functions in Reason ?)

ArthurGodet commented on solver/src/main/java/org/chocosolver/solver/constraints/binary/PropAbsoluteLight.java:
```diff
@@ -0,0 +1,104 @@
+                absY.updateUpperBound(u, this, lcg() ? Reason.r(Y.getMinLit(), Y.getMaxLit()) : Reason.undef());
+            } else if (u <= 0) {
+                absY.updateLowerBound(-u, this, lcg() ? Reason.r(Y.getMaxLit()) : Reason.undef());
+                absY.updateUpperBound(-l, this, lcg() ? Reason.r(Y.getMaxLit(), Y.getMinLit()) : Reason.undef());
```
Why is there `Y.getMaxLit()` in the explanation ?

cprudhom replied:
Same reason as previous comment

cprudhom committed:
Javadoc : update
+ taking PR comments into account

ArthurGodet committed:
Replace Cumulative implementation by state-of-the-art implementation of TimeTabling and OverloadChecking (#1165) and fixes #1114

ArthurGodet committed:
Update CHANGES.md

NelusBreizh committed:
Explained Bimodal AllDifferent (#1173)

jgFages committed:
IntVar no longer extends Iterable (#1177)

cprudhom committed:
LCG : fix incomplete initialisation of sparse domain lit-based variable

cprudhom committed:
MiniSat : fix ccmin_mode 1 and 2 (should be evaluated though)

cprudhom committed:
LCG : allow sorting literals on solution or failure (disabled by default)

cprudhom committed:
LCG : disabling additional clauses posting when at least one variable is bounded

cprudhom committed:
CT : add invariant check when CT* is used

jgFages committed:
Performance improvement in PropXPlusYEqZ and PropAbsolute (#1178)

jgFages committed:
Merge pull request #1179

* copy based reliable iterators for swap and list based sets
* fix linkedlist and graph var
* fix swap sets + refactoring
* add unit test
* optimize method
* add test group
* add tests in ReifiedConnectivityTest

cprudhom committed:
Fix tests related to PR #1179

cprudhom committed:
CT : change the way estimated memory for CT is computed

cprudhom committed:
Tests : increase timeOut of testReifiedConnectivity

cprudhom committed:
Flatzinc : deal with unary table constraints + expression-based setvar
